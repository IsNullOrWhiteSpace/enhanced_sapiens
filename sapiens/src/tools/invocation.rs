
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use tracing::debug;

use crate::tools::ToolInvocationInput;

/// Error while extracting tool invocations
#[derive(Debug, thiserror::Error, Clone, Serialize, Deserialize)]
pub enum InvocationError {
    /// Invalid yaml
    #[error("Invalid yaml: {0}")]
    InvalidYaml(String),
    /// No invocation found in the document
    #[error("No Action found")]
    NoInvocationFound,
    /// No valid invocation found in the document
    #[error("No valid Action found: {0}")]
    NoValidInvocationFound(String),
    /// Too many yaml blocks
    #[error("Too many ({0}) yaml blocks. Only one is expected.")]
    TooManyYamlBlocks(usize),
}

/// One of several T
#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum Invocation<T> {
    Single(T),
    Multiple(Vec<T>),
}

/// extract on or several T from a string.
///
/// `data` can be a single yaml of T or of a Vec<T> or a list of yaml
/// documents separated by `---`.
fn extract_from_yaml<T>(data: &str) -> Result<Vec<T>, InvocationError>
where
    T: DeserializeOwned,
{
    let mut invocations = vec![];

    for doc in serde_yaml::Deserializer::from_str(data) {
        // is it a list of T or a single T?
        let attempt: Result<Invocation<T>, _> = Deserialize::deserialize(doc);
        match attempt {
            Ok(invocation) => match invocation {
                Invocation::Single(t) => {
                    invocations.push(t);
                }
                Invocation::Multiple(ts) => {
                    invocations.extend(ts);
                }
            },
            Err(e) => {
                debug!(error = %e, "Failed to deserialize as a list of T or a single T");
                return Err(InvocationError::InvalidYaml(e.to_string()));
            }
        }
    }

    if invocations.is_empty() {
        Err(InvocationError::NoInvocationFound)
    } else {
        Ok(invocations)
    }
}

/// Extracted invocations
#[derive(Debug, Serialize, Deserialize)]
pub(crate) struct ExtractedInvocations {
    pub(crate) invocations: Vec<ToolInvocationInput>,
    pub(crate) yaml_block_count: usize,
}

/// Find all the invocations in a markdown document.
pub(crate) fn find_all(data: &str) -> Result<ExtractedInvocations, InvocationError> {
    let mut err: Option<InvocationError> = None;

    let mut invocations = vec![];
    let mut yaml_block_count = 0;

    let mut lines = data.lines();

    while let Some(line) = lines.next() {
        if line.trim().is_empty() {
            continue;
        }

        // we have start of a yaml block
        if line.trim().starts_with("```yaml") {
            // collect the lines until the end of the yaml block
            let mut yaml = vec![];

            for line in lines.by_ref() {
                if line.trim().starts_with("```") {
                    break;
                }

                yaml.push(line);
            }

            // put them together
            let yaml = yaml.join("\n");

            yaml_block_count += 1;

            // does that make valid invocations?
            match extract_from_yaml(&yaml) {
                Ok(more) => {
                    invocations.extend(more);
                }
                Err(e) => {
                    // debug!(error = %e, "Failed to extract invocation from yaml");

                    err = Some(e);
                }
            }
        }
    }

    if invocations.is_empty() {
        if let Some(err) = err {
            Err(err)
        } else {
            Err(InvocationError::NoInvocationFound)
        }
    } else {
        Ok(ExtractedInvocations {
            invocations,
            yaml_block_count,
        })
    }
}

#[cfg(test)]
mod tests {
    use indoc::indoc;
    use insta::{assert_display_snapshot, assert_yaml_snapshot};
    use serde_yaml::Number;

    #[tokio::test]
    async fn test_find_invocations_one_yaml() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    ```        
    Some other text
    "#};

        let invocations = super::find_all(data).unwrap();

        assert_eq!(invocations.invocations.len(), 1);
    }

    #[tokio::test]
    async fn test_find_multiple_invocations() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search1
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    ---
    tool_name: Search2
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    ```        
    Some other text
    ```yaml
    tool_name: Search3
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    something: else
    ```        
    Some other text
    ```yaml
    - tool_name: Search4
      parameters:
        q: Marcel Deneuve
        excluded_terms: Resident Evil
        num_results: 10
    - tool_name: Search5
      parameters:
        q: Marcel Deneuve
        excluded_terms: Resident Evil
        num_results: 10
    ```                
    Some other text
    "#};

        let invocations = super::find_all(data).unwrap();

        assert_eq!(invocations.invocations.len(), 5);

        assert_yaml_snapshot!(invocations);
    }

    #[tokio::test]
    async fn test_extraction_of_one_yaml() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    ```        
    Some other text
    "#};

        let tool_invocations = super::find_all(data).unwrap();

        assert_eq!(tool_invocations.invocations.len(), 1);

        let invocation = &tool_invocations.invocations[0];

        assert_eq!(invocation.tool_name, "Search");
    }

    #[tokio::test]
    async fn test_extraction_of_one_yaml_with_output() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    output: 
      something: | 
        Marcel Deneuve is a character in the Resident Evil film series, playing a minor role in Resident Evil: Apocalypse and a much larger role in Resident Evil: Extinction. Explore historical records and family tree profiles about Marcel Deneuve on MyHeritage, the world's largest family network.
    ```        
    Some other text
    "#};

        let tool_invocations = super::find_all(data).unwrap();

        assert_eq!(tool_invocations.invocations.len(), 1);

        let invocation = &tool_invocations.invocations[0];

        assert_eq!(invocation.tool_name, "Search");
        assert_eq!(invocation.parameters.get("q").unwrap(), "Marcel Deneuve");
        assert_eq!(
            invocation.parameters.get("excluded_terms").unwrap(),
            "Resident Evil"
        );
        assert_eq!(
            invocation.parameters.get("num_results").unwrap(),
            &serde_yaml::Value::Number(Number::from(10))
        );
        assert!(!invocation.junk.is_empty());
        assert!(invocation.junk.get("output").is_some());
    }

    #[tokio::test]
    async fn test_extraction_of_three_yaml_with_output() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search1
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    output: 
      something: | 
        Marcel Deneuve is a character in the Resident Evil film series, playing a minor role in Resident Evil: Apocalypse and a much larger role in Resident Evil: Extinction. Explore historical records and family tree profiles about Marcel Deneuve on MyHeritage, the world's largest family network.
    ```        
    Some other text
    ```yaml
    tool_name: Search2
    parameters:
      q: Marcel Prouse
      excluded_terms: La Recherche du Temps Perdu
      num_results: 10
    ```        
    Some other other text
    ```yaml
    tool_name: Search3
    parameters:
      q: Marcel et son Orchestre
      excluded_terms: Les Vaches
      num_results: 10
    ```
    That's all folks!          
    "#};

        let tool_invocations = super::find_all(data).unwrap();

        assert_eq!(tool_invocations.invocations.len(), 3);

        let invocation = &tool_invocations.invocations[0];
        assert_eq!(invocation.tool_name, "Search1");