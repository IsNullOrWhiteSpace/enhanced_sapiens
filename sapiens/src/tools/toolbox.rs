
use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;
use tracing::{debug, info};

use crate::tools;
use crate::tools::invocation::InvocationError;
use crate::tools::{
    AdvancedTool, TerminalTool, TerminationMessage, Tool, ToolDescription, ToolUseError,
};