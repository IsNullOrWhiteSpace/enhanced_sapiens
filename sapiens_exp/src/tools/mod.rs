
//! Bake the acceptance test for the task in the tools
//! Scenario -> (Recipe, Config)
//!
//! Recipe -> (Task, Acceptance criteria)
//!
//! Acceptance criteria -> Tools
//!
//! run_until_completion: (Task, Config, Tools) -> (Trace, Stats, Acceptance
//! results)
//!
//! evaluate::Trial::analyze: (Trace, Stats, Acceptance results) -> Analysis
//!
//! evaluate::Trial::build: (Trace, State, Acceptance results, Task, Analysis)
//! -> Trial

use std::marker::PhantomData;
use std::ops::DerefMut;
use std::sync::Arc;

use sapiens::tools::{Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError};
use serde::{Deserialize, Serialize};
use tokio::sync::Mutex;

/// Scenario 0: Preparing a cereal bowl
pub mod scenario_0;

mod scenario_1;

/// The state of a scenario
pub trait State: Send + Sync {
    /// Reset the state to its initial state
    fn reset(&mut self);
