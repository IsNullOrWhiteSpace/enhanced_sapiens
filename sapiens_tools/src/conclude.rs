
use std::fmt::Debug;

use sapiens::tools::{
    Describe, ProtoToolDescribe, ProtoToolInvoke, TerminalTool, TerminationMessage,
    ToolDescription, ToolUseError,
};
use sapiens_derive::{Describe, ProtoToolDescribe, ProtoToolInvoke};
use serde::{Deserialize, Serialize};
use tokio::sync::Mutex;

/// A tool to conclude a task.
/// You have to use this to once you have the answer to the task with your
/// conclusion.
#[derive(Default, ProtoToolDescribe, ProtoToolInvoke)]
#[tool(
    name = "Conclude",
    input = "ConcludeToolInput",
    output = "ConcludeToolOutput"
)]
pub struct ConcludeTool {
    done: Mutex<Option<ConcludeToolInput>>,
}

impl Debug for ConcludeTool {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ConcludeTool").finish()
    }
}

#[async_trait::async_trait]
impl TerminalTool for ConcludeTool {
    async fn is_done(&self) -> bool {
        // lock
        let done = self.done.lock().await;
        done.is_some()
    }

    async fn take_done(&self) -> Option<TerminationMessage> {
        // lock
        {
            let mut done = self.done.lock().await;
            done.take().map(|input| input.into())
        }
    }
}

/// A tool that is called to wrap the task.
#[derive(Debug, Clone, Serialize, Deserialize, Describe)]
pub struct ConcludeToolInput {
    /// The final answer for this task. Plain text ONLY. No string interpolation
    /// supported. MANDATORY. Call directly from `SandboxPython` Tool for long
    /// answers.
    pub conclusion: String,
    /// The original question that was asked to the user. No string
    /// interpolation supported, only plain text. MANDATORY.
    pub original_question: String,
}

impl From<ConcludeToolInput> for TerminationMessage {
    fn from(input: ConcludeToolInput) -> Self {
        Self {
            conclusion: input.conclusion,
            original_question: input.original_question,
        }
    }
}

/// ConcludeToolOutput - empty
#[derive(Serialize, Deserialize, Describe)]
pub struct ConcludeToolOutput {}

impl ConcludeTool {
    #[tracing::instrument(skip(self))]
    async fn invoke_typed(
        &self,
        input: &ConcludeToolInput,
    ) -> Result<ConcludeToolOutput, ToolUseError> {
        // lock
        {
            let mut done = self.done.lock().await;

            if done.is_some() {
                return Err(ToolUseError::InvocationFailed(
                    "This task is already done.".to_string(),
                ));
            }

            // set done
            *done = Some(input.clone());
        }

        Ok(ConcludeToolOutput {})
    }
}