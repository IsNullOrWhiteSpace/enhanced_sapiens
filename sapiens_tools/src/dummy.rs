
use std::fmt::Debug;

use sapiens::tools::{Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError};
use sapiens_derive::{Describe, ProtoToolDescribe, ProtoToolInvoke};
use serde::{Deserialize, Serialize};

/// A tool that is called to test stuffs
#[derive(Debug, Default, ProtoToolDescribe, ProtoToolInvoke)]
#[tool(name = "Dummy", input = "DummyToolInput", output = "DummyToolOutput")]
pub struct DummyTool {}

/// A tool that is called to test stuffs
#[derive(Debug, Serialize, Deserialize, Describe)]
pub struct DummyToolInput {
    /// Well. MANDATORY.
    pub blah: String,
}

/// DummyToolOutput not very significant
#[derive(Serialize, Deserialize, Describe)]
pub struct DummyToolOutput {
    /// Not much.
    pub something: String,
}

impl DummyTool {
    #[tracing::instrument(skip(self))]
    async fn invoke_typed(&self, input: &DummyToolInput) -> Result<DummyToolOutput, ToolUseError> {
        Ok(DummyToolOutput {
            something: input.blah.clone() + " and something else",
        })
    }
}