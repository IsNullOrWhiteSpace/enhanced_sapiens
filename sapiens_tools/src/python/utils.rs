
use pyo3::types::{IntoPyDict, PyDict, PyFloat, PyList, PyTuple};
use pyo3::{PyObject, Python, ToPyObject};
use sapiens::tools::{FieldFormat, ToolDescription};
use serde::{Deserialize, Serialize};
use serde_yaml::Value;

#[derive(thiserror::Error, Debug)]
pub(crate) enum PyConversionError {
    #[error("Invalid conversion: {error}")]
    InvalidConversion { error: String },
    #[error("dict key not serializable: {typename}")]
    DictKeyNotSerializable { typename: String },
    #[error("Invalid cast: {typename}")]
    InvalidCast { typename: String },
}

// inspired from https://github.com/mozilla-services/python-canonicaljson-rs/blob/62599b246055a1c8a78e5777acdfe0fd594be3d8/src/lib.rs#L87-L167
#[allow(clippy::redundant_closure_call)]
pub(crate) fn to_yaml(py: Python, obj: &PyObject) -> Result<Value, PyConversionError> {
    macro_rules! return_cast {
        ($t:ty, $f:expr) => {
            if let Ok(val) = obj.downcast::<$t>(py) {
                return $f(val);
            }
        };
    }

    macro_rules! return_to_value {
        ($t:ty) => {
            if let Ok(val) = obj.extract::<$t>(py) {
                return serde_yaml::to_value(val).map_err(|error| {
                    PyConversionError::InvalidConversion {
                        error: format!("{}", error),
                    }
                });
            }
        };
    }

    if obj.is_none(py) {
        return Ok(Value::Null);
    }

    return_to_value!(String);
    return_to_value!(bool);
    return_to_value!(u64);
    return_to_value!(i64);

    return_cast!(PyDict, |x: &PyDict| {
        let mut map = serde_yaml::Mapping::new();
        for (key_obj, value) in x.iter() {
            let key = if key_obj.is_none() {
                Ok("null".to_string())
            } else if let Ok(val) = key_obj.extract::<bool>() {
                Ok(if val {
                    "true".to_string()
                } else {
                    "false".to_string()
                })
            } else if let Ok(val) = key_obj.str() {
                Ok(val.to_string())
            } else {
                Err(PyConversionError::DictKeyNotSerializable {
                    typename: key_obj
                        .to_object(py)
                        .as_ref(py)
                        .get_type()
                        .name()
                        .map(|x| x.to_string())
                        .unwrap_or_else(|_| "unknown".to_string()),
                })
            };
            map.insert(Value::String(key?), to_yaml(py, &value.to_object(py))?);
        }
        Ok(Value::Mapping(map))
    });

    return_cast!(PyList, |x: &PyList| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyTuple, |x: &PyTuple| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyFloat, |x: &PyFloat| {
        Ok(Value::Number(serde_yaml::Number::from(x.value())))
    });

    // At this point we can't cast it, set up the error object
    Err(PyConversionError::InvalidCast {
        typename: obj
            .as_ref(py)
            .get_type()
            .name()
            .map(|x| x.to_string())
            .unwrap_or_else(|_| "unknown".to_string()),
    })
}

pub(crate) fn value_to_object(val: Value, py: Python<'_>) -> PyObject {
    match val {
        Value::Null => py.None(),
        Value::Bool(x) => x.to_object(py),
        Value::Number(x) => {
            let oi64 = x.as_i64().map(|i| i.to_object(py));
            let ou64 = x.as_u64().map(|i| i.to_object(py));
            let of64 = x.as_f64().map(|i| i.to_object(py));
            oi64.or(ou64).or(of64).expect("number too large")
        }
        Value::String(x) => x.to_object(py),
        Value::Sequence(x) => {
            let inner: Vec<_> = x.into_iter().map(|x| value_to_object(x, py)).collect();
            inner.to_object(py)
        }
        Value::Mapping(x) => {
            let iter = x
                .into_iter()
                .map(|(k, v)| (value_to_object(k, py), value_to_object(v, py)));
            IntoPyDict::into_py_dict(iter, py).into()
        }
        Value::Tagged(_) => panic!("tagged values are not supported"),
    }
}

/// Format of a field in a tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimpleFormat {
    /// Name of the field
    pub name: String,
    /// Type of the field
    pub r#type: String,
    /// Whether the field is optional
    pub optional: bool,
    /// Description of the field
    pub description: String,
}

impl From<FieldFormat> for SimpleFormat {
    fn from(part: FieldFormat) -> Self {
        SimpleFormat {
            name: part.name,
            r#type: part.r#type,
            optional: part.optional,
            description: part.description,
        }
    }
}

impl ToPyObject for SimpleFormat {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.to_object(py)).unwrap();
        dict.set_item("type", self.r#type.to_object(py)).unwrap();
        dict.set_item("optional", self.optional.to_object(py))
            .unwrap();
        dict.set_item("description", self.description.to_object(py))
            .unwrap();
        dict.into()
    }
}

/// A simplified version of the tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub(crate) struct SimpleToolDescription {
    /// Name of the tool
    pub name: String,
    /// Description of the tool
    pub description: String,
    /// Input format of the tool
    pub parameters: Vec<SimpleFormat>,
    /// Output format of the tool
    pub responses_content: Vec<SimpleFormat>,
}

impl From<ToolDescription> for SimpleToolDescription {
    fn from(desc: ToolDescription) -> Self {
        let parameters = desc
            .parameters
            .fields
            .into_iter()
            .map(|x| x.into())
            .collect();
        let responses_content = desc
            .responses_content
            .fields
            .into_iter()
            .map(|x| x.into())
            .collect();

        SimpleToolDescription {
            name: desc.name,
            description: desc.description,
            parameters,
            responses_content,
        }
    }
}

impl ToPyObject for SimpleToolDescription {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.clone()).unwrap();
        dict.set_item("description", self.description.clone())
            .unwrap();
        dict.set_item("parameters", self.parameters.clone())
            .unwrap();
        dict.set_item("responses_content", self.responses_content.clone())
            .unwrap();
        dict.into()
    }
}