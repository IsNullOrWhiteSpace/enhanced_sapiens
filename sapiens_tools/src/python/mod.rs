
use std::cmp::Ordering;
use std::collections::HashMap;

use convert_case::{Case, Casing};
use pyo3::indoc::{formatdoc, indoc};
use pyo3::prelude::*;
use pyo3::types::{IntoPyDict, PyDict};
use sapiens::tools::toolbox::{invoke_simple_from_toolbox, Toolbox};
use sapiens::tools::{
    AdvancedTool, Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError,
};
use sapiens_derive::{Describe, ProtoToolDescribe};
use serde::{Deserialize, Serialize};
use serde_yaml::Value;
use tracing::trace;

/// Conversion tools
pub(crate) mod utils;

use crate::python::utils::SimpleToolDescription;

const MAX_OUTPUT_SIZE: usize = 512;

// FUTURE(ssoudan) install pySWIP

/// A tool that runs sandboxed Python code. Use this to transform data.
///
/// - To use another Tool with parameters `input_field_1` and `input_field_2`
///   and result fields `output_field_1` and `output_field_2` use:
/// ```python
/// result = tools.ToolName(input_field_1=..., input_field_2=...)
/// print(result['output_field_1'])
/// print(result['output_field_2'])
/// ```
/// - Only stdout and stderr are captured and made available (limited to 512B
///   total). If the output is larger, use `tools.Conclude` directly from the
///   code.
/// - List available tools with `tools.list()`. And returns a list of
///   `{'name':.., 'description':.., 'parameters':.., 'responses_content':..,
///   }`.
/// - `open`|`exec` are forbidden.
/// - Limited libraries available: urllib3, requests, sympy, numpy,
/// BeautifulSoup4, feedparser, arxiv.
/// - No PIP.
#[derive(Debug, Default, ProtoToolDescribe)]
#[tool(
    name = "SandboxedPython",
    input = "PythonToolInput",
    output = "PythonToolOutput"
)]
pub struct PythonTool {}

/// The input of the Python tool
#[derive(Debug, Serialize, Deserialize, Describe)]
pub struct PythonToolInput {
    /// The Python code to run. MANDATORY
    pub code: String,
}

/// The output of the Python tool
#[derive(Serialize, Deserialize, Describe)]
pub struct PythonToolOutput {
    /// The stdout output of the Python code.
    pub stdout: String,
    /// The stderr output of the Python code.
    pub stderr: String,
}

#[pyclass]
#[derive(Default)]
struct Logging {
    output: String,
}

#[pymethods]
impl Logging {
    fn write(&mut self, data: &str) {
        self.output.push_str(data);
    }
}

#[pyclass(unsendable)]
struct ToolsWrapper {
    toolbox: Toolbox,
    tool_list: Vec<SimpleToolDescription>,
}

impl ToolsWrapper {
    async fn new(toolbox: Toolbox) -> Self {
        let tools = toolbox.describe().await;
        let tool_list = tools
            .into_values()
            .map(SimpleToolDescription::from)
            .collect::<Vec<_>>();

        ToolsWrapper { toolbox, tool_list }
    }
}

#[pymethods]
impl ToolsWrapper {
    // list all tools
    #[pyo3(signature = ())]
    fn list(&self, py: Python<'_>) -> PyResult<PyObject> {
        let tools = self.tool_list.to_object(py);
        Ok(tools)
    }

    // invoke a tool
    #[pyo3(signature = (tool_name, input))]
    fn invoke(
        &self,
        py: Python<'_>,
        tool_name: &str,
        input: Option<&PyDict>,
    ) -> PyResult<PyObject> {
        // convert PyDict to a serde_yaml::Value
        let input = if let Some(input) = input {
            let input: PyObject = input.into();

            utils::to_yaml(py, &input).map_err(|e| {
                pyo3::exceptions::PyException::new_err(format!("Invalid input: {}", e))
            })?
        } else {
            Value::default()
        };

        let (tx, mut rx) = tokio::sync::oneshot::channel::<Result<Value, ToolUseError>>();

        // release the GIL to allow the thread to run
        py.allow_threads(move || {