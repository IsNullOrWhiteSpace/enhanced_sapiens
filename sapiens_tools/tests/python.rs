
use indoc::indoc;
use insta::assert_display_snapshot;
use pyo3::PyResult;
use sapiens::tools::invocation::InvocationError;
use sapiens::tools::toolbox::{invoke_tool, InvokeResult, Toolbox};
use sapiens_tools::conclude::ConcludeTool;
use sapiens_tools::dummy::DummyTool;
use sapiens_tools::python::PythonTool;

#[pyo3_asyncio::tokio::test]
async fn test_tool_invocation() -> PyResult<()> {
    let data = indoc! {r#"
    # Action
    ```yaml        
    tool_name: SandboxedPython
    parameters:
        code: |
            print("Hello world!")          
    ```
    "#};

    let mut toolbox = Toolbox::default();
    toolbox.add_advanced_tool(PythonTool::default()).await;

    let res = invoke_tool(toolbox, data).await;

    match res {
        InvokeResult::Success {
            tool_name, result, ..
        } => {
            assert_eq!(tool_name, "SandboxedPython");
            assert_eq!(result, "stdout: |\n  Hello world!\nstderr: ''\n");
        }
        _ => panic!("Unexpected response: {:?}", res),
    }

    Ok(())
}

#[pyo3_asyncio::tokio::test]
async fn test_tool_simple_invocation() -> PyResult<()> {
    let data = indoc! {r#"
    # Action
    ```yaml        
    tool_name: Conclude
    parameters:
        original_question: |
            print("Hello world!")
        conclusion: |
            Hello world!          
    ```
    "#};

    let mut toolbox = Toolbox::default();
    toolbox.add_advanced_tool(PythonTool::default()).await;
    toolbox.add_tool(ConcludeTool::default()).await;

    let res = invoke_tool(toolbox, data).await;

    match res {
        InvokeResult::Success {
            tool_name, result, ..
        } => {
            assert_eq!(tool_name, "Conclude");
            assert_display_snapshot!(result);
        }
        _ => panic!("Unexpected response: {:?}", res),
    }

    Ok(())
}

#[pyo3_asyncio::tokio::test]
async fn test_tool_invocation_in_python() -> PyResult<()> {
    let data = indoc! {r#"
    # Action
    ```yaml        
    tool_name: SandboxedPython
    parameters:
        code: |
            print("Hello world!")
            rooms = toolbox.invoke("Dummy", {"blah": "blah"})
            print(rooms)
            rooms = tools.Dummy(blah="blah")
            print(rooms)          
    ```
    "#};

    let mut toolbox = Toolbox::default();
    toolbox.add_advanced_tool(PythonTool::default()).await;
    toolbox.add_tool(DummyTool::default()).await;

    let res = invoke_tool(toolbox, data).await;

    match res {
        InvokeResult::Success {
            tool_name, result, ..
        } => {
            assert_eq!(tool_name, "SandboxedPython");
            assert_display_snapshot!(result);
        }
        _ => panic!("Unexpected response: {:?}", res),
    }

    Ok(())
}

#[pyo3_asyncio::tokio::test]
async fn test_exit_in_python() -> PyResult<()> {
    let data = indoc! {r#"
    # Action
    ```yaml        
    tool_name: SandboxedPython
    parameters:
        code: |
            print("Bye bye!")
            exit(0)
    ```
    "#};

    let mut toolbox = Toolbox::default();
    toolbox.add_advanced_tool(PythonTool::default()).await;
    toolbox.add_tool(DummyTool::default()).await;

    let res = invoke_tool(toolbox, data).await;

    match res {
        InvokeResult::Error { tool_name, e, .. } => {
            assert_eq!(tool_name, "SandboxedPython");
            assert_display_snapshot!(e);
        }
        _ => panic!("Unexpected response: {:?}", res),
    }

    Ok(())
}

#[pyo3_asyncio::tokio::test]
async fn test_multiple_tool_invocations() -> PyResult<()> {
    let data = indoc! {r#"
    # Action
    ```yaml        
    tool_name: SandboxedPython
    parameters:
        code: |
            print("Hello world 1!")          
    ```
    
    # And another action
    ```yaml        
    tool_name: SandboxedPython
    parameters:
        code: |
            print("Hello world 2!")          
    ```
    
    # And yet another action
    ```        
    tool_name: SandboxedPython
    input:
        code: |
            print("Hello world 3!")          
    ```
    "#};

    let mut toolbox = Toolbox::default();
    toolbox.add_advanced_tool(PythonTool::default()).await;

    let res = invoke_tool(toolbox, data).await;

    match res {
        InvokeResult::NoValidInvocationsFound {
            e: InvocationError::TooManyYamlBlocks(2),
            invocation_count: 2,
        } => {
            // This is expected
        }
        _ => panic!("Unexpected response: {:?}", res),
    }

    Ok(())
}

#[pyo3_asyncio::tokio::test]
async fn test_python() -> PyResult<()> {
    let mut toolbox = Toolbox::default();
    toolbox.add_tool(DummyTool::default()).await;
    toolbox.add_advanced_tool(PythonTool::default()).await;

    let data = indoc! {r#"```yaml
   tool_name: SandboxedPython
   parameters:
     code: |           
       args = {
           'blah': "hello"
       }
       output = tools.Dummy(**args)           
      
       something = output['something']                       

       print(f"And the result is: {something}")
   ```
"#};

    let res = invoke_tool(toolbox, data).await;

    match res {
        InvokeResult::Success {
            tool_name, result, ..
        } => {
            assert_eq!(tool_name, "SandboxedPython");
            assert_display_snapshot!(result);
        }
        _ => panic!("Unexpected response: {:?}", res),
    }

    Ok(())
}

#[pyo3_asyncio::tokio::test]
async fn test_python_docstring() -> PyResult<()> {
    let mut toolbox = Toolbox::default();
    toolbox.add_tool(DummyTool::default()).await;
    toolbox.add_advanced_tool(PythonTool::default()).await;

    let data = indoc! {r#"```yaml
   tool_name: SandboxedPython
   parameters:
     code: |                  
       output = help(tools.Dummy)           
       print(f"And the docstring is: {output}")
   ```
"#};

    let res = invoke_tool(toolbox, data).await;

    match res {
        InvokeResult::Success {
            tool_name, result, ..
        } => {
            assert_eq!(tool_name, "SandboxedPython");
            assert_display_snapshot!(result);
        }
        _ => panic!("Unexpected response: {:?}", res),
    }

    Ok(())
}