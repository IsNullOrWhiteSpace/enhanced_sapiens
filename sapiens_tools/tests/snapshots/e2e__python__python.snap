---
source: sapiens_tools/tests/python.rs
expression: output
---
stdout: |
  And the result is: hello and something else
stderr: ''

